lua/scratch/actor.lua

```diff
@@ -0,0 +1,130 @@
+local utils = require("scratch.utils")
+-- local MANUAL_INPUT_OPTION = "MANUAL_INPUT"
+
+---@class Scratch.Actor
+---@field base_dir string
+---@field win_config vim.api.keyset.win_config @see: nvim_open_win() {config}
+---@field filetypes string[]
+---@field manual_text string
+---@field file_picker? "fzflua" | "telescope"
+---@field filetype_details Scratch.FiletypeDetails
+---@field localKeys Scratch.LocalKeyConfig[]
+local M = {}
+
+---@class Scratch.ActionOpts
+---@field content? string[] content will be put into the scratch file
+
+---@alias Scratch.Action fun(ft: string, opts?: Scratch.ActionOpts): nil
+
+---@param filename string
+---@param win_conf? vim.api.keyset.win_config
+---@param content? string[]
+---@param local_keys? Scratch.LocalKeyConfig
+---@param cursor? Scratch.Cursor
+function M:scratchByName(filename, win_conf, content, local_keys, cursor)
+  local abs_path = self.base_dir .. filename
+  local fto = {}
+  for i in filename:gmatch("([^%.]+)") do
+    table.insert(fto, i)
+  end
+  local ft = fto[#fto]
+  content = content or self.filetype_details[ft] and self.filetype_details[ft].content
+  cursor = cursor or self.filetype_details[ft] and self.filetype_details[ft].cursor
+  utils.scratch(abs_path, win_conf or self.win_config, content, local_keys, cursor)
+end
+
+---@param ft string
+---@param win_conf? vim.api.keyset.win_config
+---@param content? string[]
+function M:scratchByType(ft, win_conf, content, local_keys, cursor)
+  local abs_path = self.base_dir .. utils.gen_filename(ft)
+  content = content or self.filetype_details[ft] and self.filetype_details[ft].content
+  cursor = cursor or self.filetype_details[ft] and self.filetype_details[ft].cursor
+  utils.scratch(abs_path, win_conf or self.win_config, content, local_keys, cursor)
+end
+
+---@return string[]
+function M:get_all_filetypes()
+  local combined_filetypes = {}
+  local cash = {}
```

Comment:
Instead of sneaking through all combined_filetypes each time we append an item, just caching with ~O(1)